File: intro-fp/src/main/scala/com/kpler/scala/kata/introduction/Bomberman.scala
package com.kpler.scala.kata.introduction

import java.io.{BufferedWriter, FileWriter}
import scala.annotation.tailrec
import scala.io.Source
import scala.util.Using

object Bomberman {


  case class Point(x: Int, y: Int)

  case class Board(width: Int, height: Int, content: List[String]) {

    /*
     Return p plus the adjacent list of points to p at the coordinates:
     - (p.x - 1, y)
     - (p.x + 1, y)
     - (p.x, p.y - 1)
     - (p.x, p.y + 1)
    The coordinates must be valid:
     - 0 <= x < width
     - 0 <= y < height
    p itself must be part of the returned list

    hints: Use a for comprehension

    time: 8 minutes
    */
    private def neighbors(p: Point): List[Point] = ???

    /*
     Return the points inside the board where a bomb is planted
     A bomb is defined by the 'O' character

    hints: You can get a value and its index from an iterable with 'zipWithIndex'
    example:
      List('a', 'b', 'c').zipWithIndex => List( ('a', 0), ('b', 1), ('c', 2) )
    A String is an iterator
      "I love Scala".zipWithIndex => Vector(
      (I,0), ( ,1), (l,2), (o,3), (v,4), (e,5), ( ,6), (S,7), (c,8), (a,9), (l,10), (a,11)
      )

    time: 5 minutes
    */
    def getBombs: List[Point] = ???

    /*
      Returns the new board after the bombs located at 'bombLocations' have detonated

      Each cell inside the board at the coordinates of a bomb and all its neighbors become an empty cell.
      An empty cell is defined by the '.' character

      hints:
      Re-use the nested internal function seen earlier

      You might want to temporarily convert a row (String) of the board into a StringBuffer
      so that you can replace one character with insert
        StringBuffer("I love Scala").insert(0, 'i') => "i love Scala"

      You can access an element of a list with '()'
        val lst = List(4, 5, 6)
        lst(1) => 5

      You cannot modify a list directly (immutability) but build a new updated list with 'updated'
        List(4, 5, 6).updated(0, 10) => List(10, 5, 6)

     time: 12 minutes
     */
    def detonate(bombLocations: List[Point]): Board = {
      val destroyed = bombLocations.flatMap(this.neighbors)

      @tailrec
      def internal(b: List[Point], newContent: List[StringBuffer]): List[String] = b match {
        case Nil => newContent.map(_.toString)
        case destroyed :: rest =>
          val line = newContent(destroyed.y).insert(destroyed.x, '.')
          internal(rest, newContent.updated(destroyed.y, line))
      }

      val updatedContent = internal(destroyed, content.map(StringBuffer(_)))
      Board(this.width, this.height, updatedContent)
    }

    /*
     Returns a new board with bombs inside every cell
     A bomb is defined by the 'O' character

     hints:
     You have to replace each string in content with a new string filled with 'O'
     You might use the 'map' function on both 'List' and 'String'

     time: 6 minutes
     */
    def fullOfBombs(): Board = ???
  }


  def bomberman(nbSeconds: Int, board: Board): Board = nbSeconds match {
    case 0 | 1 => board
    case _ if nbSeconds % 2 == 0 => board.fullOfBombs()
    case _ if nbSeconds % 4 == 3 => board.fullOfBombs().detonate(board.getBombs)
    case _ => board

  }

  @main def bomberman_main(args: String*): Unit = {
    Using(Source.fromFile("bomberman-input.txt")) { source =>
      val allLines = source.getLines.toList
      val firstLine = allLines.head.strip().split(" ")
      val r = firstLine(0).toInt
      val c = firstLine(1).toInt
      val n = firstLine(2).toInt

      val result = bomberman(n, Board(c, r, allLines.tail))

      Using(new BufferedWriter(new FileWriter("bomberman-output-scala.txt"))) { writer =>
        writer.write(result.content.mkString("\n"))
      }


    }
  }
}

File: intro-fp/src/main/python/com/kpler/scala/kata/introduction/bomberman.py
#!/bin/python3

#
# Complete the 'bomberMan' function below.
#
# The function is expected to return a STRING_ARRAY.
# The function accepts following parameters:
#  1. INTEGER n
#  2. STRING_ARRAY grid
#

Board = list[str]
Point = tuple[int, int]


def neighbors(p: Point, width: int, height: int) -> list[Point]:
    (x, y) = p
    return [(i, y) for i in [x - 1, x, x + 1] if 0 <= i < width] + \
        [(x, j) for j in [y - 1, y + 1] if 0 <= j < height]


def detonate(bombs: list[Point], board: Board) -> Board:
    height = len(board)
    width = len(board[0])
    board_copy = board.copy()
    all_exploded_cells = {(i, j) for bomb in bombs for (i, j) in neighbors(bomb, width, height)}
    for (i, j) in all_exploded_cells:
        row: str = board_copy[j]
        board_copy[j] = row[:i] + '.' + row[(i + 1):]
    return board_copy


def get_bombs(board: Board) -> list[Point]:
    height = len(board)
    width = len(board[0])
    return [(i, j) for i in range(width) for j in range(height) if board[j][i] == 'O']


def bomber_man(nb_seconds: int, board: Board) -> Board:
    height = len(board)
    width = len(board[0])
    bombs = get_bombs(board)
    full_of_bombs: Board = ['O' * width] * height

    board_after_first_explode = detonate(bombs, full_of_bombs)

    """
    0: board
    1: board
    2: full_of_bombs
    3: board_after_first_explode (3 % 4 == 3)
    4: full_of_bombs
    5: other_explode_state (5 % 4 == 1)
    6: full_of_bombs
    7: board_after_first_explode (7 % 4 == 3)
    8: full_of_bombs
    9: other_explode_state (9 % 4 == 1)
    """

    if nb_seconds == 0 or nb_seconds == 1:
        return board
    else:
        if nb_seconds % 2 == 0:
            return full_of_bombs
        elif nb_seconds % 4 == 3:
            return board_after_first_explode
        else:
            return board


if __name__ == '__main__':
    with open('bomberman-input.txt', mode='r') as input_file:
        all_lines = input_file.readlines()
        first_line = all_lines[0].rstrip().split()

        r = int(first_line[0])
        c = int(first_line[1])
        n = int(first_line[2])

        grid = [all_lines[i] for i in range(1, r)]

        result = bomber_man(n, grid)

        with open('bomberman-output.txt', mode='w') as output_file:
            output_file.write('\n'.join(result))
            output_file.write('\n')
